configuration_broker/parser/parser.cc

```diff
@@ -0,0 +1,192 @@
+
+#include <compartment.h>
+#include <cstdlib>
+#include <debug.hh>
+#include <string.h>
+#include <thread.h>
+
+#include <magic_enum/magic_enum.hpp>
+
+// Expose debugging features unconditionally for this compartment.
+using Debug = ConditionalDebug<true, "Parser">;
+
+#include "parser.h"
+#include "parser_helper.h"
+
+// Set for Items we are allowed to register a parser for
+#include "../config_broker/config_broker.h"
+
+#include "../rgb_led/rgb_led.h"
+#define RGB_LED_CONFIG "rgb_led"
+DEFINE_PARSER_CONFIG_CAPABILITY(RGB_LED_CONFIG, sizeof(RGB_LED_Config), 1800);
+
+#include "../user_led/user_led.h"
+#define USER_LED_CONFIG "user_led"
+DEFINE_PARSER_CONFIG_CAPABILITY(USER_LED_CONFIG, sizeof(User_LED_Config), 1800);
+
+#include "../logger/logger.h"
+#define LOGGER_CONFIG "logger"
+DEFINE_PARSER_CONFIG_CAPABILITY(LOGGER_CONFIG, sizeof(LoggerConfig), 500);
+
+//
+// Register all of our parsers. This needs to be run before
+// any values can be accepted.
+//
+// There is no init mechanism in cheriot and threads are not
+// expected to terminate, so rather than have a separate thread
+// just to run this which then blocks we expose it as method for
+// the Broker to call when the first item is published.
+//
+void __cheri_compartment("parser") parserInit()
+{
+	// RGB LED Config Parser
+	if (set_parser(PARSER_CONFIG_CAPABILITY(RGB_LED_CONFIG),
+	               parse_RGB_LED_config) < 0)
+	{
+		Debug::log("Failed to register parser for rgb led");
+	}
+
+	// USER LED Config Parser
+	if (set_parser(PARSER_CONFIG_CAPABILITY(USER_LED_CONFIG),
+	           parse_User_LED_config) < 0)
+	{
+		Debug::log("Failed to register parser for user led");
+	}
+
+	// Logger Config Parser
+	if (set_parser(PARSER_CONFIG_CAPABILITY(LOGGER_CONFIG),
+	               parse_logger_config) < 0)
+	{
+		Debug::log("Failed to register parser for logger");
+	}
+}
+
+// /////////////////////////////////////////////
+//
+// Parsers for specific configuration items
+//
+// Note that we run in a sandbox compartment to
+// contain the blast radius of any parsing errors.
+//
+// /////////////////////////////////////////////
+
+//
+// Parse a json string into a LoggerConfig object.
+//
+int __cheri_callback parse_logger_config(const char *json, void *dst)
+{
+	auto        *config = static_cast<LoggerConfig *>(dst);
+	JSONStatus_t result;
+
+	// Check we have valid JSON
+	result = JSON_Validate(json, strlen(json));
+	if (result != JSONSuccess)
+	{
+		Debug::log("thread {} Invalid JSON {}", thread_id_get(), json);
+		return -1;
+	}
+
+	// query the individual values and populate the logger config object
+	bool parsed = true;
+	parsed = parsed && get_string(json, "host.address", config->host.address);
+	parsed =
+	  parsed && get_number<uint16_t>(json, "host.port", &config->host.port);
+	parsed = parsed && get_enum<logLevel>(json, "level", &config->level);
+
+	// Free any heap the parser might have left allocated
+	auto heap_freed = heap_free_all(MALLOC_CAPABILITY);
```

Comment:
I thought this compartment wasn't allowed to allocate or free memory?

Reply:
Ideally it wouldn't - but as per the comment in parser.h using magic_enum to convert from string values in JSON requires new(), so the next best thing seems to be to force a tidy up if it left anything around (from what can see it never does).   I think it makes a good story about using compartments to isolate code that its not always possible to clean up first.  

In the next version I'm going to move each parser into its own compartment and give them each a very small heap quota so that even if one parser did get implemented without this it wouldn't affect any the parsing of any other items.